NBitcoin/Secp256k1/Musig/MusigPrivNonce.cs

```diff
@@ -0,0 +1,195 @@
+﻿#if HAS_SPAN
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace NBitcoin.Secp256k1.Musig
+{
+
+#if SECP256K1_LIB
+	public
+#endif
+	class MusigPrivNonce : IDisposable
+	{
+		readonly static RandomNumberGenerator rand = new RNGCryptoServiceProvider();
+
+
+		/// <summary>
+		/// This function derives a secret nonce that will be required for signing and
+		/// creates a private nonce whose public part intended to be sent to other signers.
+		/// </summary>
+		/// <param name="context">The context</param>
+		/// <param name="sessionId32">A unique session_id32. It is a "number used once". If empty, it will be randomly generated.</param>
+		/// <param name="signingKey">Provide the message to be signed to increase misuse-resistance. If you do provide a signingKey, sessionId32 can instead be a counter (that must never repeat!). However, it is recommended to always choose session_id32 uniformly at random. Can be null.</param>
+		/// <param name="msg32">Provide the message to be signed to increase misuse-resistance. Can be empty.</param>
+		/// <param name="combinedKey">Provide the message to be signed to increase misuse-resistance. Can be null.</param>
+		/// <param name="extraInput32">Provide the message to be signed to increase misuse-resistance. The extra_input32 argument can be used to provide additional data that does not repeat in normal scenarios, such as the current time. Can be empty.</param>
+		/// <returns>A private nonce whose public part intended to be sent to other signers</returns>
+		public static MusigPrivNonce GenerateMusigNonce(
+						   Context? context,
+						   ReadOnlySpan<byte> sessionId32,
+						   ECPrivKey? signingKey,
+						   ReadOnlySpan<byte> msg32,
+						   ECXOnlyPubKey? combinedKey,
+						   ReadOnlySpan<byte> extraInput32)
+		{
+			if (!(sessionId32.Length is 32) && !(sessionId32.Length is 0))
+				throw new ArgumentException(nameof(sessionId32), "sessionId32 must be 32 bytes or 0 bytes");
+			if (sessionId32.Length is 0)
+			{
+				var bytes = new byte[32];
+				sessionId32 = bytes.AsSpan();
+				rand.GetBytes(bytes);
+			}
+			if (!(msg32.Length is 32) && !(msg32.Length is 0))
+				throw new ArgumentException(nameof(msg32), "msg32 must be 32 bytes or 0 bytes");
+			if (!(extraInput32.Length is 32) && !(extraInput32.Length is 0))
+				throw new ArgumentException(nameof(extraInput32), "extraInput32 must be 32 bytes or 0 bytes");
+
+			Span<byte> key32 = stackalloc byte[32];
+			if (signingKey is null)
+			{
+				key32 = key32.Slice(0, 0);
```

Comment:
? I don't understand what `Slice(0, 0)` does here.

Reply:
mmmh maybe I should better used nullable Span if possible.

Reply:
Can't use nullable on spans